src/mpid/common/iqueue/mpidu_iqueue.c

```diff
@@ -0,0 +1,190 @@
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_POOL_ALLOC);
+    return mpi_errno;
+  fn_fail:
+    if (new_pool) {
+        MPL_free(new_pool);
+    }
+    if (slab) {
+        free_fn(slab);
+    }
+    goto fn_exit;
+}
+
+void MPIDU_iqueue_pool_free(MPIDU_iqueue_pool_t * pool)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_IQUEUE_POOL_FREE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_IQUEUE_POOL_FREE);
+
+    if (pool->slab) {
+        pool->free_fn(pool->slab);
+    }
+    MPL_free(pool);
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_POOL_FREE);
+}
+
+void MPIDU_iqueue_queue_alloc(MPIDU_iqueue_pool_t * pool, int index, MPIDU_iqueue_queue_t ** queue)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_IQUEUE_QUEUE_ALLOC);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_IQUEUE_QUEUE_ALLOC);
+
+    /* static allocation, get queue by index */
+    MPIR_Assert(index < pool->max_num_queues);
+    MPIDU_iqueue_queue_t *new_queue = pool->all_queues + sizeof(MPIDU_iqueue_queue_t) * index;
+    new_queue->u.handle = MPIDU_IQUEUE_POOL_GET_OBJ_HANDLE(pool, new_queue);
+    new_queue->u.head = NULL;
+    MPL_atomic_relaxed_store_ptr(&new_queue->u.tail, NULL);
+    *queue = new_queue;
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_QUEUE_ALLOC);
+}
+
+void MPIDU_iqueue_queue_free(MPIDU_iqueue_queue_t * queue)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_IQUEUE_QUEUE_FREE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_IQUEUE_QUEUE_FREE);
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_QUEUE_FREE);
+    return;
+}
+
+static void iqueue_cellblock_alloc(MPIDU_iqueue_pool_t * pool)
+{
+    int index = MPL_atomic_fetch_add_int(&pool->cellblk_mgr->next_free_idx, 1);
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_ALLOC);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_ALLOC);
+
+    if (index < pool->max_num_cellblocks) {
+        pool->cells = (MPIDU_iqueue_cell_t *) ((uintptr_t) pool->all_cells
+                                               + index * pool->num_cells_in_block
+                                               * pool->cell_size);
+        for (int i = 0; i < pool->num_cells_in_block; i++) {
+            MPIDU_iqueue_cell_t *cell = MPIDU_IQUEUE_GET_CELL_BY_INDEX(pool, index);
+            cell->next = NULL;
+            cell->prev = 0;
+            MPL_atomic_relaxed_store_int(&cell->in_use, 0);
+        }
+    } else {
+        MPL_atomic_fetch_sub_int(&pool->cellblk_mgr->next_free_idx, 1);
+    }
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_ALLOC);
+}
+
+static void iqueue_cellblock_free(MPIDU_iqueue_pool_t * pool, MPIDU_iqueue_cell_t * cells)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_FREE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_FREE);
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_FREE);
+    return;
+}
+
+void MPIDU_iqueue_cell_reserve(MPIDU_iqueue_pool_t * pool)
```

Comment:
This initializes all the cells in the pool so `cells_init` is a more fitting name.